File: src/android/app/src/main/java/com/crickettechnology/trees/QtRect.java
package com.crickettechnology.trees;

class QtRect
{
    public QtRect(QtSpan latitude, QtSpan longitude)
    {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public QtSpan latitude;
    public QtSpan longitude;

    public static final QtRect invalid = new QtRect(QtSpan.invalid, QtSpan.invalid);

    public double getArea()
    {
        return this.latitude.getLength() * this.longitude.getLength();
    }

    public QtRect getQuarter(int i, int j)
    {
        return new QtRect(this.latitude.getHalf(i), this.longitude.getHalf(j));
    }

    public QtRect inflated(double ratio)
    {
        return new QtRect(this.latitude.inflated(ratio),
                          this.longitude.inflated(ratio));
    }

    boolean contains(QtLocation location)
    {
        return this.latitude.contains(location.latitude) &&
               this.longitude.contains(location.longitude);
    }

    boolean contains(QtRect rect)
    {
        return this.latitude.contains(rect.latitude) &&
               this.longitude.contains(rect.longitude);
    }

    boolean intersects(QtRect rect)
    {
        return !(this.latitude.max <= rect.latitude.min ||
                 this.latitude.min >= rect.latitude.max ||
                 this.longitude.max <= rect.longitude.min ||
                 this.longitude.min >= rect.longitude.max);
    }
}

File: src/ios/trees/QtRect.swift
import Foundation
import GoogleMaps

struct QtRect
{
    init(latitude: QtSpan, longitude: QtSpan)
    {
        self.latitude = latitude
        self.longitude = longitude
    }

    init(location loc0: QtLocation, location loc1: QtLocation)
    {
        let latSpan = QtSpan(min: min(loc0.latitude, loc1.latitude), max:max(loc0.latitude, loc1.latitude))
        let lngSpan = QtSpan(min: min(loc0.longitude, loc1.longitude), max:max(loc0.longitude, loc1.longitude))
        self.init(latitude: latSpan, longitude: lngSpan)
    }

    var latitude: QtSpan
    var longitude: QtSpan

    static let invalid = QtRect(latitude: QtSpan.invalid, longitude: QtSpan.invalid)

    var area: QtCoordinate
    {
        get { return self.latitude.length * self.longitude.length }
    }

    func getCorner(_ i: Int,_ j: Int) -> QtLocation
    {
        return QtLocation(latitude: self.latitude.getEnd(i), longitude: self.longitude.getEnd(j))
    }

    func getQuarter(_ i: Int,_ j: Int) -> QtRect
    {
        return QtRect(latitude: self.latitude.getHalf(i), longitude: self.longitude.getHalf(j))
    }

    func inflated(ratio: QtCoordinate) -> QtRect
    {
        return QtRect(latitude:self.latitude.inflated(ratio:ratio),
                      longitude:self.longitude.inflated(ratio:ratio))
    }

    func contains(location: QtLocation) -> Bool
    {
        return self.latitude.contains(coord:location.latitude) &&
               self.longitude.contains(coord:location.longitude)
    }

    func contains(rect: QtRect) -> Bool
    {
        return self.latitude.contains(span:rect.latitude) &&
               self.longitude.contains(span:rect.longitude)
    }

    func intersects(rect: QtRect) -> Bool
    {
        return !(self.latitude.max <= rect.latitude.min ||
                 self.latitude.min >= rect.latitude.max ||
                 self.longitude.max <= rect.longitude.min ||
                 self.longitude.min >= rect.longitude.max)
    }
}


////////////////////////////////////////
// GMSCoordinateBounds conversions

extension QtRect
{
    init(_ rect: GMSCoordinateBounds) 
    {
        self.init(location: QtLocation(rect.northEast), location: QtLocation(rect.southWest))
    }
}

extension GMSCoordinateBounds
{
    convenience init(_ rect: QtRect)
    {
        let northeast = CLLocationCoordinate2D(rect.getCorner(1,0))
        let southwest = CLLocationCoordinate2D(rect.getCorner(0,1))
        self.init(coordinate: northeast, coordinate: southwest)
    }
}
